File: ios/CounterWidget/DatabaseService.swift
//
//  DatabaseService.swift
//  CounterWidgetExtension
//
//  Created by tsinis on 14.08.2022.
//

import Foundation
import SQLite3

// Database Service for SQLite database, very similar to Dart's one:
// https://github.com/tsinis/flutter_widgetkit/blob/main/lib/database_service.dart
//
// Dart is using underscore for private fields, Swift is using private keyword.
struct DatabaseService {
    // Few constants to use in the database, the same as
    // in the Dart version of this class.
    private let queryStatementInt = "SELECT value FROM counter;"
    private let appGroupId = "group.com.example.flutterWidgetkit"
    private let dbPath: String = "database.db"

    // SQLite database to use, equivalent to Dart's Database `_db` field.
    private var db: OpaquePointer?

    // Convient Swift's method to initialize the class.
    init() {
        db = openDb()
    }

    // The same method as in the Dart class, reads a value from the database,
    // with a raw SQL query and checks existence of the "count" value. In the
    // Swift version, we are returning String since we are showing it as Text
    // on the home screen widget, but in Dart, we are returning integer since
    // we are increasing it with taps on the button.
    func getCount() -> String? {
        var maybeCount: String?
        var queryStatement: OpaquePointer?
        let sqlState = sqlite3_prepare_v2(db, queryStatementInt, -1, &queryStatement, nil)
        if sqlState == SQLITE_OK {
            while sqlite3_step(queryStatement) == SQLITE_ROW {
                let value = sqlite3_column_int(queryStatement, 0)
                maybeCount = String(describing: value)
            }
        } else {
            print("SELECT statement could not be prepared")
        }
        sqlite3_finalize(queryStatement)

        return maybeCount
    }

    // The same method as in Dart class, also opens a database from app group
    // directory and checks the existence of the database. On the Dart side
    // we are also creating the database if it's not exist (on the first run).
    private func openDb() -> OpaquePointer? {
        let fileManager = FileManager.default
        let directory = fileManager.containerURL(forSecurityApplicationGroupIdentifier: appGroupId)
        let dbFile = directory!.appendingPathComponent(dbPath)
        var db: OpaquePointer?

        if sqlite3_open(dbFile.path, &db) != SQLITE_OK {
            print("Error opening database")

            return nil
        } else {
            print("Successfully opened connection to database at \(dbPath)")

            return db
        }
    }
}

File: lib/database_service.dart
import 'package:app_group_directory/app_group_directory.dart';
import 'package:flutter_widgetkit/flutter_widgetkit.dart' show WidgetKit;
// ignore: depend_on_referenced_packages
import 'package:path/path.dart' show join;
import 'package:sqflite/sqflite.dart';

/// Database Service for SQLite database, very similar to Swift's one:
/// https://github.com/tsinis/flutter_widgetkit/blob/main/ios/CounterWidget/DatabaseService.swift
///
/// Dart is using underscore for private fields, Swift is using private keyword.
class DatabaseService {
  /// Private constructor to prevent instantiation without calling
  /// async `openDb()` since Dart doesn't have async factories/constructors.
  const DatabaseService._(this._db);

  /// SQLite database to use, equivalent to Swift's OpaquePointer `db` field.
  final Database _db;

  /// Few constants to use in the database, the same as
  /// in Swift version of this class.
  static const _appGroupId = 'group.com.example.flutterWidgetkit';
  static const _dbPath = 'database.db';
  static const _queryStatementInt = "SELECT value FROM counter;";

  /// Those are not referenced in the Swift version of this class,
  /// because they are used for database creation and updating.
  static const _id = 'id';
  static const _table = 'counter';
  static const _value = 'value';

  /// Same method as in Swift class, reads a value from the database,
  /// with a raw SQL query and checking existence of the "count" value.
  /// In Swift version we are returning String since we are showing it
  /// as Text on home-screen widget, but in Dart we are returning integer
  /// since we are increasing it with a taps on the button.
  Future<int?> getCount() async {
    final map = await _db.rawQuery(_queryStatementInt);
    if (map.isEmpty) return null;
    final maybeCount = map.first[_value];

    return maybeCount is int ? maybeCount : null;
  }

  /// Same method as in Swift class, also opens a database from app group
  /// directory and checking existence of the database. On the Dart side
  /// we are also creating the database if it's not exists (on the first run).
  static Future<DatabaseService> openDb() async {
    final directory = await AppGroupDirectory.getAppGroupDirectory(_appGroupId);
    if (directory == null) throw Exception('App Group $_appGroupId not found!');
    final database = await openDatabase(
      join(directory.path, _dbPath),
      version: 1,
      onCreate: (db, _) => db.execute(
        '''
          CREATE TABLE $_table(
            $_id TEXT PRIMARY KEY,
            $_value INTEGER
          )
        ''',
      ),
    );

    return DatabaseService._(database);
  }

  /// This method is only provided on Dart side, since Swift's home-screen
  /// widget is in "read-only" mode and doesn't allow to update the count.
  Future<void> updateCount(int count) async {
    /// Insert new count to the database, if exists - just replaces it.
    await _db.insert(
      _table,
      {_id: _id, _value: count},
      conflictAlgorithm: ConflictAlgorithm.replace,
    );

    /// This will trigger a rebuild of the Swift's home-screen widget.
    return WidgetKit.reloadAllTimelines();
  }
}
